app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php

```diff
@@ -0,0 +1,196 @@
+{
+    /**
+     * @var RequestInterface
+     */
+    private $request;
+
+    /**
+     * @var CheckoutSession
+     */
+    private $checkoutSession;
+
+    /**
+     * @var ProductRepositoryInterface
+     */
+    private $productRepository;
+
+    /**
+     * @var Cart
+     */
+    private $cart;
+
+    /**
+     * @var PageFactory
+     */
+    private $resultPageFactory;
+
+    /**
+     * @var RedirectFactory
+     */
+    private $resultRedirectFactory;
+    
+    /**
+     * @var CouponFactory
+     */
+    private $couponFactory;
+    
+    /**
+     * @var Usage
+     */
+    private $couponUsage;
+    
+    /**
+     * @var ManagerInterface
+     */
+    private $messageManager;
+
+    /**
+     * @param Context $context
+     * @param CheckoutSession $checkoutSession
+     * @param ProductRepositoryInterface $productRepository
+     * @param Cart $cart
+     * @param PageFactory $resultPageFactory
+     * @param RedirectFactory $resultRedirectFactory
+     * @param CouponFactory $couponFactory
+     * @param Usage $couponUsage
+     * @param ManagerInterface $messageManager
+     */
+    public function __construct(
+        Context $context,
+        CheckoutSession $checkoutSession,
+        ProductRepositoryInterface $productRepository,
+        Cart $cart,
+        PageFactory $resultPageFactory,
+        RedirectFactory $resultRedirectFactory,
+        CouponFactory $couponFactory,
+        Usage $couponUsage,
+        ManagerInterface $messageManager
+    ) {
+        $this->request = $context->getRequest();
+        $this->checkoutSession = $checkoutSession;
+        $this->productRepository = $productRepository;
+        $this->cart = $cart;
+        $this->resultPageFactory = $resultPageFactory;
+        $this->resultRedirectFactory = $resultRedirectFactory;
+        $this->couponFactory = $couponFactory;
+        $this->couponUsage = $couponUsage;
+        $this->messageManager = $messageManager;
+    }
+
+    /**
+     * Execute action based on request and return result
+     *
+     * @return \Magento\Framework\Controller\ResultInterface
+     */
+    public function execute()
+    {
+        // Get products parameter
+        $productsParam = $this->request->getParam('products', '');
+        $couponCode = $this->request->getParam('coupon', '');
+        
+        // Clear the cart first (required by Meta spec)
+        $this->cart->truncate();
+        
+        // Parse products parameter
+        if (!empty($productsParam)) {
+            $productItems = $this->parseProductsParam($productsParam);
+            
+            // Add products to cart
+            foreach ($productItems as $item) {
+                try {
+                    $productId = $item['product_id'];
+                    $qty = $item['qty'];
+                    
+                    $product = $this->productRepository->getById($productId);
+                    $this->cart->addProduct($product, ['qty' => $qty]);
```

Comment:
I guess we can pass `$product->getId()`, instead of passing `$product` as `addProduct` method requires either `int` or `\Magento\Catalog\Model\Product` parameter